### DISCUSSION TITLE
N+1 with multiple many to one relationship on single pivot table

### DISCUSSION THREAD

**Discussion by korematter** [filamentphp/filament#5806] [Help] [answered]
Hi there,

I seem to running into an N+1 situation in Relations Manager trying to display a pivot attribute that's a belongsTo relationship accessed via a many to many relationship. 

Heres the setup. I have a many to many relationship with three tables but only really care about many-to-many relationship between two tables (styles and dances)  with the third table (competitions) just as a reference lookup. 

**styles**
id
name

**dances**
id
name

**competitions**
id
name

**dance_style**
style_id
dance_id
competition_id

`class Style extends Model
{
    use HasFactory;

    public function dances()
    {
        return $this->belongsToMany(Dance::class)->using(DanceStylePivot::class)->withPivot('competition_id');
    }
    
}`

`class DanceStylePivot extends Pivot
{
    public function competition()
    {
        return $this->belongsTo(Competition::class, 'competition_id');
    }
}`

In Relations Managers, I'm rendering the table accessing the competitions.name using pivot.competition.name

`<?php

namespace App\Filament\Resources\StyleResource\RelationManagers;

use Filament\Forms;
use Filament\Tables;
use Filament\Resources\Form;
use Filament\Resources\Table;
use Filament\Tables\Actions\AttachAction;
use Filament\Resources\RelationManagers\RelationManager;

class DancesRelationManager extends RelationManager
{
    protected static string $relationship = 'dances';

    protected static ?string $recordTitleAttribute = 'name';

    public static function form(Form $form): Form
    {
        return $form
            ->schema([
                Forms\Components\TextInput::make('name')
                    ->required()
                    ->maxLength(255),
            ]);
    }

    public static function table(Table $table): Table
    {
        return $table
            ->columns([
                Tables\Columns\TextColumn::make('pivot.competition.name'),
                Tables\Columns\TextColumn::make('name'),
            ])
            ->filters([
                //
            ])
            ->headerActions([
                    AttachAction::make()
                    ->preloadRecordSelect(),
            ])
            ->actions([
                Tables\Actions\DetachAction::make(),
            ])
            ->bulkActions([
                Tables\Actions\DeleteBulkAction::make(),
            ]);
    } 
    
}
`

This seems to be generating n+1 queries on the competitions table.
<img width="737" alt="n+1_competitions" src="https://user-images.githubusercontent.com/58536534/221735349-15d871a5-90e2-4b50-87f7-cf489fd49491.PNG">

Is there a way to eager load this table or is there another way to retrieve this third table name field?  Been trying to figure this out for days with no avail. Has anyone run into a similar issue?

Thank you! 
Steve

**Comment by danharrin** [comment, marked as answer]
So Filament will attempt to eagle load relationships. If you have a `pivot.competition.name` column, it will add a `with('pivot.competition')` to your table query automatically. If Eloquent doesn't support eagre loading through a pivot relationship, then that's probably where the problem is here. Maybe try https://github.com/ajcastro/eager-load-pivot-relations ?

**Reply by korematter** [reply]
Thanks Dan! You're right. Looks like an Eloquent issue. Also, didn't realize someone developed a package for this. Very cool. I'll try it. Alternatively, I changed the use case to make this pivot table (dance_style) to a Model so all the _id fields are a belongsTo relationship. Thanks again!
